Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs

```diff
@@ -0,0 +1,187 @@
+                    nodes[counter].SetCumulativeWeight(parent.GetCumulativeWeight() + weightSelector(items[nodes[counter].GetLevel()]));
+                    nodes[counter].SetCumulativeValue(parent.GetCumulativeValue() + valueSelector(items[nodes[counter].GetLevel()]));
+
+                    // IF cumulative weight is smaller than the weight capacity of the knapsack AND
+                    // current cumulative value is larger then the current maxCumulativeValue, update the maxCumulativeValue
+                    if (nodes[counter].GetCumulativeWeight() <= capacity && nodes[counter].GetCumulativeValue() > maxCumulativeValue)
+                    {
+                        maxCumulativeValue = nodes[counter].GetCumulativeValue();
+                        lastNodeOfOptimalPath = nodes[counter];
+                    }
+
+                    // find upperBound of this node
+                    nodes[counter].SetUpperBound(ComputeUpperBound(nodes[counter], items, capacity, weightSelector, valueSelector));
+
+                    // IF upperBound of this node is larger than maxCumulativeValue,
+                    // the current path is still possible to reach or surpass the maximum value,
+                    // add current node to nodesQueue so that nodes can be further contructed below it
+                    if (nodes[counter].GetUpperBound() > maxCumulativeValue && nodes[counter].GetCumulativeWeight() < capacity)
+                    {
+                        nodesQueue.Enqueue(nodes[counter]);
+                    }
+
+                    // repeat everything for the second node but the node's level's item is not taken
+                    counter++;
+                    nodes[counter].SetParent(parent);
+
+                    nodes[counter].SetCumulativeWeight(parent.GetCumulativeWeight());
+                    nodes[counter].SetCumulativeValue(parent.GetCumulativeValue());
+
+                    nodes[counter].SetUpperBound(ComputeUpperBound(nodes[counter], items, capacity, weightSelector, valueSelector));
+
+                    if (nodes[counter].GetUpperBound() > maxCumulativeValue)
+                    {
+                        nodesQueue.Enqueue(nodes[counter]);
+                    }
+
+                    counter++;
+                }
+            }
+
+            return GetOptimalItems(items, lastNodeOfOptimalPath);
+        }
+
+        // determine items taken based on the path that gives maximum value
+        private static T[] GetOptimalItems(T[] items, BranchAndBoundNode lastNodeOfOptimalPath)
+        {
+            List<T> takenItems = new List<T>();
+
+            BranchAndBoundNode? currentNode = lastNodeOfOptimalPath;
+            int numberOfNodes = lastNodeOfOptimalPath.GetLevel();
+
+            for (int i = numberOfNodes; i >= 0; i--)
+            {
+                // the node's level's item is taken, add to knapsack taken items list
+                if(currentNode != null)
```

Comment:
This line doesn't make sense to me, please review working with nulls and return type of GetParent()

Reply:
It doesn't make sense to me as well, the problem with the "non-nullable parent node must contain a non-null value when exiting a constructor" error mentioned above was not encounter when I implement this algorithm in Java. 
![image](https://user-images.githubusercontent.com/86560855/127609871-6547ba7a-2893-456e-bd32-651fde0817b4.png)
![image](https://user-images.githubusercontent.com/86560855/127609666-0f6655a6-0efb-4788-8e41-840469785a61.png)
Looking at the first picture, although the currentNode is never null, but because is it nullable, I have to add an if statement to check if it is null. Second picture shows the original implementation, it was using a while loop, but due to the same reason I have changed it to a for loop which loops for the same amount of times and produces the same result.

Reply:
I've updated this too